apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -0,0 +1,427 @@
+    Actions = [{state_timeout, 0, health_check}],
+    {next_state, connecting, Data, Actions};
+% Connecting state health_check timeouts.
+% First clause supports desired behavior on initial connection.
+handle_event(state_timeout, health_check, connecting, #data{status = disconnected} = Data) ->
+    {next_state, disconnected, Data};
+handle_event(state_timeout, health_check, connecting, Data) ->
+    connecting_health_check(Data);
+%% The connected state is entered after a successful start of the callback mod
+%% and successful health_checks
+handle_event(enter, _OldState, connected, Data) ->
+    Actions = [{state_timeout, ?HEALTHCHECK_INTERVAL, health_check}],
+    {next_state, connected, Data, Actions};
+handle_event(state_timeout, health_check, connected, Data) ->
+    perform_connected_health_check(Data);
+% Disconnected state entered when a healtcheck has failed.
+handle_event(enter, _OldState, disconnected, #data{id = InstId} = Data) ->
+    UpdatedData = Data#data{status = disconnected},
+    ets:delete(?ETS_TABLE, InstId),
+    {next_state, disconnected, UpdatedData}.
+
+%%------------------------------------------------------------------------------
+%% internal functions
+%%------------------------------------------------------------------------------
+
+handle_connecting_state_enter_event(Data) ->
+    Result = emqx_resource:call_start(Data#data.id, Data#data.mod, Data#data.config),
+    case Result of
+        {ok, ResourceState} ->
+            UpdatedData = Data#data{state = ResourceState, status = connecting},
+            %% Perform an initial health_check immediately before transitioning into a connected state
+            Actions = [{state_timeout, 0, health_check}],
+            {next_state, connecting, UpdatedData, Actions};
+        {error, Reason} ->
+            %% Keep track of the error reason why the connection did not work
+            %% so that the Reason can be returned when the verification call is made.
+            UpdatedData = Data#data{status = disconnected, error = Reason},
+            Actions = [{state_timeout, 0, health_check}],
+            {next_state, connecting, UpdatedData, Actions}
+    end.
+
+handle_remove_event(From, ClearMetrics, Data) ->
+    do_stop(Data),
+    ets:delete(?ETS_TABLE, Data#data.id),
+    case ClearMetrics of
+        true -> ok = emqx_metrics_worker:clear_metrics(resource_metrics, Data#data.id);
+        false -> ok
+    end,
+    {stop_and_reply, normal, [{reply, From, ok}]}.
+
+do_start(InstId, Group, ResourceType, Config, Opts) ->
+    % The state machine will make the actual call to the callback/resource module after init
+    _ = emqx_resource_manager_sup:start_child(InstId, Group, ResourceType, Config, Opts),
+    ok = emqx_metrics_worker:create_metrics(
+        resource_metrics,
+        InstId,
+        [matched, success, failed, exception],
+        [matched]
+    ),
+    wait_for_resource_ready(InstId, maps:get(wait_for_resource_ready, Opts, 5000)),
```

Comment:
Q: should we check for timeout here?

Reply:
I initially did that here. We have some built in assumptions in a bunch of places that this succeeds...or at least that it returns an `{ok State}` tuple rather than returning any sort of error. And then the status of the connection is checked or even in a bunch of cases it isn't checked. There seems to be an assumption that it just worked or that other parts of the logic will handle what happens if it doesn't work. So I changed it to how the current implementation behaves to try and keep as close to the same 'contract' as possible.

As an example there is this in `emqx_authn_jwt.erl`:
```
create2(
    #{
        use_jwks := true,
        verify_claims := VerifyClaims
    } = Config
) ->
    ResourceId = emqx_authn_utils:make_resource_id(?MODULE),
    {ok, _} = emqx_resource:create_local(
        ResourceId,
        ?RESOURCE_GROUP,
        emqx_authn_jwks_connector,
        connector_opts(Config)
    ),
    {ok, #{
        jwk_resource => ResourceId,
        verify_claims => VerifyClaims
    }}.
```
If I changed the behavior so it could return an error would we just ignore it here? We do that in `emqx_authn_mongodb.erl`:
```
    _ = emqx_resource:create_local(
        ResourceId,
        ?RESOURCE_GROUP,
        emqx_connector_mongo,
        Config,
        #{}
    ),
    {ok, NState}.
```
Or since we now provide an expanded potential return compared to the current implementation do we need to refactor how all of the `authn` and `authz` modules work so they can properly respond/react to the additional potential return values?

Making a different choice might be the correct answer, and making these changes might actually be the best thing in the long run...but depending on the choice made it could mean an additional ~20 files need changing in at least some small part and that seemed like a lot to try and include in this PR.